dsl/flatten.cc

```diff
@@ -0,0 +1,304 @@
+                if (send != nullptr && send->fun == core::Names::sig()) {
+                    exprs[i - 1].staticLevel = expr.staticLevel;
+                }
+            }
+        }
+
+        // these will store the bodies of the `class << self` blocks we create at the end
+        vector<ast::ClassDef::RHS_store> nestedBlocks;
+        for (int level = 2; level <= highestLevel; level++) {
+            nestedBlocks.emplace_back();
+        }
+
+        // this vector contains all the possible RHS target locations that we might move to
+        vector<ast::ClassDef::RHS_store *> targets;
+        // 0 and 1 both go into the class itself
+        targets.emplace_back(&rhs);
+        targets.emplace_back(&rhs);
+        // 2 and up go into the to-be-created `class << self` blocks
+        for (auto &tgt : nestedBlocks) {
+            targets.emplace_back(&tgt);
+        }
+
+        // move everything to its appropriate target
+        for (auto &expr : exprs) {
+            if (auto methodDef = ast::cast_tree<ast::MethodDef>(expr.expr.get())) {
+                methodDef->setIsSelf(expr.staticLevel > 0);
+            }
+            targets[expr.staticLevel]->emplace_back(std::move(expr.expr));
+        }
+
+        // generate the nested `class << self` blocks as needed and add them to the class
+        for (auto &body : nestedBlocks) {
+            auto classDef =
+                ast::MK::Class(loc, loc,
+                               make_unique<ast::UnresolvedIdent>(core::Loc::none(), ast::UnresolvedIdent::Class,
+                                                                 core::Names::singleton()),
+                               {}, std::move(body), ast::ClassDefKind::Class);
+            rhs.emplace_back(std::move(classDef));
+        }
+
+        return rhs;
+    }
+
+public:
+    FlattenWalk() {
+        newMethodSet();
+    }
+    ~FlattenWalk() {
+        ENFORCE(methodScopes.empty());
+    }
+
+    unique_ptr<ast::ClassDef> preTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        newMethodSet();
+        return classDef;
+    }
+
+    unique_ptr<ast::Expression> postTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDef->rhs = addMethods(ctx, std::move(classDef->rhs), classDef->loc);
+        return classDef;
+    };
+
+    // We might want to move sends as well: either if they're method modifiers like `private` or `protected` or if
+    // they're `sig`s. If so, then we'll treat them like we treat methods on our method stack
+    unique_ptr<ast::Send> preTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        auto &methods = curMethodSet();
+        if (methods.stack.size() == 0) {
+            return send;
+        }
+
+        if (send->fun == core::Names::sig()) {
+            methods.stack.emplace_back(methods.methods.size(), 0, send.get());
+            methods.methods.emplace_back();
```

Comment:
it's not clear for me why are we adding a default initialized datastructure here. 
Can we simplify the datastructures from the current semantic of  `vector<stack<>>`
Maybe I'm missing something but it's not clear for me what does the 2-dimensional structure model

Reply:
I've refactored in a way that removed the default-initialized vector here.